File: leetCode/c/p7_ReverseInteger/lc7.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

int reverse(int x){

    int sol = 0, tmp = 0;
    bool neg = false;

    if(x < 0) {
        neg = true;
        x *= -1;
    }

    while(x > 0) {
        sol = sol * 10 + (x % 10);
        x /= 10; 
    }

    if(neg)
        sol *= -1;

    return sol; 
}

/* Test */

void testFunction(int x, int sol, int testNumber) {
       
    int tmp; 
    double timeSpent;
     
    clock_t tic = clock();
    
    tmp = reverse(x); 
    
    clock_t toc = clock(); 
    
    timeSpent = (double) (toc - tic) * 1000000 / CLOCKS_PER_SEC;
 
    printf("\nTEST%d: ", testNumber);
 
    if(tmp == sol) 
        printf("PASSED IN %.2fus\n", timeSpent);
    else  
        printf("FAILED\n");
}

int main(void) {

    testFunction(123, 321, 1);    
    testFunction(-123, -321, 2);
    testFunction(120, 21, 3);
    testFunction(0, 0, 4);

    return EXIT_SUCCESS;
}

File: leetCode/python/p7_ReverseInteger/lc7.py
#!/usr/bin/python3

import time

class Solution:
    """
    Given a 32-bit signed integer, reverse digits of an integer.
    """
    def reverse(self, x: int) -> int:   
        sign = False  
        
        if x < 0:
            sign = True
            x = -x 
        x_str = str(x)[::-1]
        sol = int(x_str) 
        
        if sign:
            sol *= -1  
        index = 0 
         
        # remove zeros from begining
        if x_str != '0': 
            while x_str[index] == '0':
                x_str = x_str[:index] + x_str[index+1:] 
        
        return sol 
    
    def test_function(self, x: int, sol: int, test_num: int):
        start_time = time.time() 
        tmp = self.reverse(x) 
        end_time = time.time()
        if tmp == sol:
            print('\nTEST' + str(test_num) + ': PASSED IN ' + '{:.2f}us'.format((end_time - start_time) * 1000000))            
        else:
            print('\nTEST FAILED') 
         
# Test 1
            
num = 123
sol = Solution()
sol.test_function(num, 321, 1)

# Test 2

num = -123
sol = Solution()
sol.test_function(num, -321, 1)

# Test 3

num = 120
sol = Solution()
sol.test_function(num, 21, 1)

# Test 4

num = 0
sol = Solution()
sol.test_function(num, 0, 1)
